gzip/gzip.go

```diff
@@ -40,25 +68,42 @@ func (c CompressionCodec) Code() int8 {
 func (c CompressionCodec) Encode(src []byte) ([]byte, error) {
 	buf := bytes.Buffer{}
 	buf.Grow(len(src)) // guess a size to avoid repeat allocations.
-	writer := gzip.NewWriter(&buf)
+	writer := writerPool.Get().(*gzip.Writer)
+	writer.Reset(&buf)
 
 	_, err := writer.Write(src)
 	if err != nil {
+		// don't return writer to pool on error.
 		return nil, err
 	}
 
+	// note that the gzip reader must be closed in order for it to write
+	// out trailing contents.  Flush is insufficient.  it is okay to re-use
+	// the writer even after it's closed by Resetting it.
 	err = writer.Close()
 	if err != nil {
+		// don't return writer to pool on error.
 		return nil, err
 	}
+
+	writerPool.Put(writer)
+
 	return buf.Bytes(), err
 }
 
 // Decode implements the kafka.CompressionCodec interface.
 func (c CompressionCodec) Decode(src []byte) ([]byte, error) {
-	reader, err := gzip.NewReader(bytes.NewReader(src))
-	if err != nil {
-		return nil, err
+	reader := readerPool.Get().(*gzip.Reader)
+	res, err := func() ([]byte, error) {
```

Comment:
While clever I feel like this construct is a pit hard to read, is it only preventing us from repeating a `readerPool.Put` call?

Reply:
Thanks for catching...that's some cruft from a previous iteration.  Since this is copy/paste work things got out of sync.